rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,815 @@
+            * tf.math.cumprod(
+                1 - mask_combined_sequence_sentence,
+                axis=1,
+                exclusive=True,
+                reverse=True,
+            )
+        )
+
+        # The new mask is used to distribute the sentence features at the sequence
+        # positions marked by 1s. The sentence features' dimensionality effectively
+        # changes from `(batch_size, 1, feature_dim)` to `(batch_size, max_seq_length+1,
+        # feature_dim)`, but the array is sparse, with real features present only at
+        # positions determined by 1s in the mask.
+        sentence_tensor = sentence_feature_positions_mask * sentence_tensor
+
+        # Padding of sequence-level features is increased by 1 in the sequence
+        # dimension to match the shape of modified sentence-level features.
+        sequence_tensor = tf.pad(sequence_tensor, [[0, 0], [0, 1], [0, 0]])
+
+        # Sequence- and sentence-level features effectively get concatenated by
+        # summing the two padded feature arrays like this (batch size  = 1):
+        # [[seq1, seq2, seq3, 0, 0]] + [[0, 0, 0, sent1, 0]] =
+        # = [[seq1, seq2, seq3, sent1, 0]]
+        return sequence_tensor + sentence_tensor
+
+    def call(
+        self,
+        inputs: Tuple[
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            tf.Tensor,
+        ],
+        training: bool = False,
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+        """Combine multiple 3-D dense/sparse feature tensors into one.
+
+        Arguments:
+            inputs: Tuple containing:
+                sequence_features: List of 3-D dense or sparse tensors with token-level
+                    features.
+                sentence_features: List of 3-D dense or sparse tensors with sentence-level
+                    features.
+                sequence_feature_lengths: Dense 1-D tensor containing the lengths of 
+                    sequence-level features for the batch.
+            training: Python boolean indicating whether the layer should behave in
+                training mode (applying dropout to sparse tensors if applicable) or in
+                inference mode (not applying dropout).
+
+        Returns:
+            combined features: a 3-D tensor containing the combined sequence- and sentence-
+                level, sparse & dense features.
+            mask_combined_sequence_sentence: a binary 3-D tensor with 1s in place of real
+                features in the combined feature array, and 0s in place of fake features.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        sequence_feature_lengths = inputs[2]
+
+        # This mask is specifically for sequence-level features.
+        mask_sequence = _compute_mask(sequence_feature_lengths)
+
+        # Process sequence-level features if any are present.
+        if self._feature_types_present[SEQUENCE]:
```

Comment:
I think, it makes sense to create helpers for these 2 `if` with `if` inside, and return None

Reply:
Ah, okay. Addresses the above comment (L452-453) too.